compiler/luci/service/src/Nodes/CircleReshape.cpp

```diff
@@ -34,4 +40,100 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleReshape *node)
+{
+
+loco::TensorShape Algorithm::visit(const luci::CircleReshape *node)
+{
+  LOGGER(l);
+
+  const loco::DataType S32 = loco::DataType::S32;
+
+  loco::TensorShape shape_by_input;
+  {
+    LUCI_ASSERT(node->shape(), "2nd input shape() should not be nullptr");
+
+    // Only support node's shape() is CircleConst with S32
+    // TODO support other node with other types
+    auto const_shape_node = dynamic_cast<luci::CircleConst *>(node->shape());
+    if (const_shape_node != nullptr)
+    {
+      LUCI_ASSERT(const_shape_node->dtype() == S32, "Only support int32 CircleConst");
+
+      shape_by_input.rank(const_shape_node->size<S32>());
+
+      for (uint32_t axis = 0; axis < shape_by_input.rank(); ++axis)
+      {
+        shape_by_input.dim(axis) = const_shape_node->at<S32>(axis);
+        if (const_shape_node->at<S32>(axis) < 0)
+          shape_by_input.dim(axis).unset();
+      }
+    }
+    else
+    {
+      // We use shape from the node itself
+      shape_by_input = own_shape(node);
+    }
+  }
+
+  loco::TensorShape shape_by_attr;
+  {
+    shape_by_attr.rank(node->newShape()->rank());
+
+    for (uint32_t axis = 0; axis < shape_by_attr.rank(); ++axis)
+    {
+      shape_by_attr.dim(axis) = node->newShape()->dim(axis);
+    }
+  }
+
+  if (!(shape_by_input == shape_by_attr))
+  {
+    INFO(l) << "CircleReshape: Two new shape information mismatched : " << std::endl;
+    INFO(l) << "   shape_by_input : " << shape_by_input << std::endl;
+    INFO(l) << "   shape_by_attr : " << shape_by_attr << std::endl;
+  }
+
+  loco::TensorShape output_shape = shape_by_input;
+
+  // One of the dimensions of input shape can have special value -1,
+  // which should not be inferred.
+  const auto input_shape = circle_shape(loco::must_cast<CircleNode *>(node->tensor()));
+
+  uint32_t input_unknown_dim_index = UINT32_MAX;
+  uint32_t input_element_count = 1;
+  for (uint32_t i = 0; i < input_shape.rank(); i++)
+  {
+    input_element_count *= (input_shape.dim(i).known() ? input_shape.dim(i).value() : 1);
+    if (!input_shape.dim(i).known())
+    {
+      if (input_unknown_dim_index != UINT32_MAX)
+        INTERNAL_EXN("More than one unknown dimension");
+      input_unknown_dim_index = i;
+    }
+  }
+
+  // One of the dimensions of output shape can have special value -1,
+  // meaning its actual value might be inferred.
+  uint32_t output_unknown_dim_index = UINT32_MAX;
+  uint32_t output_element_count = 1;
+  for (uint32_t i = 0; i < output_shape.rank(); i++)
+  {
+    output_element_count *= (output_shape.dim(i).known() ? output_shape.dim(i).value() : 1);
+    if (!output_shape.dim(i).known())
+    {
+      if (output_unknown_dim_index != UINT32_MAX)
+        INTERNAL_EXN("More than one unknown dimension");
+      output_unknown_dim_index = i;
+    }
+  }
+
+  if (output_unknown_dim_index != UINT32_MAX && input_unknown_dim_index == UINT32_MAX)
+  {
+    output_shape.dim(output_unknown_dim_index) = input_element_count / output_element_count;
```

Comment:
This line assumed that all dimensions of input tensor was static.
When any of them is dynamic/unknown, this logic becomes invalid.

Overall dynamic handling should be considered :)

Reply:
Thank you for all of your kind and detailed comments!
I've been reading them carefully and am currently working on supporting static/dynamic new shape input.
However, the problem is that I can't figure out how to determine the output shape when the new shape tensor (the second input for the reshape node) is dynamic.
For example, the following is `Net_Gather_SparseToDense_AddV2_000.tflite` from `build/compiler/common-artifacts`, which is preventing me from successfully building.
![image](https://github.com/user-attachments/assets/54fd358d-5547-43e5-bfe6-ea603a01372e)  
As you can see, the new shape input tensor (denoted as 3) indicates that the new shape is dynamic.
At runtime, it could be a tensor like `[x, y, z]`, so I think the reshape operation can be done successfully in most cases.
But how can I determine/infer the output shape at compile time?
I can't even understand why the output shape is 1 * 10 in the picture.
Could you give me some advice?

Reply:
> However, the problem is that I can't figure out how to determine the output shape when the new shape input tensor is dynamic.

I think that we may use `shape_by_attr` instead of `shape_by_input`, according to `https://github.com/tensorflow/tensorflow/blob/2fe29379f6449d263b787eb636020fd81017b8bf/tensorflow/lite/kernels/reshape.cc#L131-L143

> As you can see, the new shape input tensor (denoted as 3) indicates that the new shape is dynamic.
At runtime, it could be a tensor like [x, y, z], so I think the reshape operation can be done successfully in most cases.

If the dimension of `reshape->shape()` is static as shown in the image, we can set the output shape as you said :)
```cpp
auto input_shape_node = reshape->shape();
reshape->rank(circle_shape(reshape->shape()).dim(0).value());
for (uint32_t d = 0;d < reshape->rank();++d)
  reshape->dim(d).unset();
```

> I can't even understand why the output shape is 1 * 10 in the picture.

I think the recipe is invalid...
As the first committer of the recipe went out from out team, we may fix by ourselves :(
BTW, is that test was valid before this PR or you enabled the test...?

Reply:
Thanks for your explanation!  
I still don't fully understand how to manipulate the `shape_by_attr`, but I'll do my best to study the TFLite code and figure it out.

For now, I'm considering limiting this PR to focus only on constant shape inference, which includes:
- Dynamic output shape inference when the new shape is provided by the node's input and is constant.
- Dynamic output shape inference when the new shape is provided by the node's attribute and is constant.

The current commit covers these cases.  
This change was discussed during our SSAFY mentoring session with @zetwhite.  
I plan to implement dynamic output shape inference for non-constant inputs in a separate PR.

> BTW, is that test was valid before this PR or you enabled the test...?

The implementation I initially devised (which is not in this PR) failed to build due to the error `Optimized graph is invalid` on `Net_Gather_SparseToDense_AddV2_000.tflite`. However, the new commit builds successfully and passes all the tests because I used `own_shape` for non-constant input shapes, similar to the previous implementation. I plan to address non-constant shape input support in another PR.

Reply:
> For now, I'm considering limiting this PR to focus only on constant shape inference

If so, I will explain little bit more detailed about non-const case at another PR :)

Reply:
Nevertheless, I think that the recipe is still invalid and `own_shape` is not used... T.T